### PR TITLE
Implement printf builtin command

### PR DESCRIPTION
**Pull request by AlexandreYang**
## Summary
- Add safe `printf` builtin with all standard format specifiers (`%s`, `%b`, `%c`, `%d`, `%i`, `%o`, `%u`, `%x`, `%X`, `%e`, `%E`, `%f`, `%F`, `%g`, `%G`, `%%`), escape sequences, width/precision modifiers (including `*` from args), and format reuse for excess arguments
- Safety hardening: `%n` rejected (security risk), `-v` rejected, format reuse bounded to 10,000 iterations with context cancellation, width/precision clamped to ±10,000
- Comprehensive test suite: 80+ unit tests, 25 GNU compatibility tests, 41 pentest tests, 46 YAML scenario tests

## Test plan
- [x] `go test ./interp/builtins/printf/...` — all unit, GNU compat, and pentest tests pass
- [x] `go test ./tests/...` — all YAML scenario tests pass
- [x] `go test ./interp/...` — full interp test suite passes (no regressions)
- [x] `RSHELL_BASH_TEST=1 go test ./tests/ -run TestShellScenariosAgainstBash -timeout 120s` — bash comparison

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by AlexandreYang — COMMENTED** [review]
## Code Review: Implement printf builtin command

### Overall Assessment: **Needs fixes** (submitted as COMMENT because reviewer is PR author — would be REQUEST_CHANGES otherwise)

The printf implementation is well-structured, thoroughly tested, and security-hardened. The `%n` rejection, width/precision clamping, iteration bounding, and context cancellation are all excellent safety measures. The import allowlist additions are correct and properly justified.

However, there are several **bash compatibility bugs** that will cause the `TestShellScenariosAgainstBash` suite to fail, and missing `skip_assert_against_bash` flags on intentionally-divergent scenarios.

### Findings Summary

| # | Priority | File | Finding |
|---|----------|------|---------|
| 1 | <sub><sub>P1 Badge</sub></sub> | `interp/builtins/printf/printf.go:133` | `printf` with no args returns exit 1 instead of bash's exit 2 |
| 2 | <sub><sub>P1 Badge</sub></sub> | `interp/builtins/printf/printf.go:340-343` | Incomplete format specifiers (`%`, `%-`, `%10`) silently succeed instead of erroring (exit 1) |
| 3 | <sub><sub>P1 Badge</sub></sub> | `interp/builtins/printf/printf.go:548-551` | Unknown format specifiers (`%z`) silently succeed instead of erroring (exit 1) |
| 4 | <sub><sub>P2 Badge</sub></sub> | `tests/scenarios/cmd/printf/errors/rejected_n_specifier.yaml` | Missing `skip_assert_against_bash: true` — bash accepts `%n` |
| 5 | <sub><sub>P2 Badge</sub></sub> | `tests/scenarios/cmd/printf/basic/no_args.yaml` | Missing `skip_assert_against_bash: true` — bash returns exit 2, not 1 |
| 6 | <sub><sub>P2 Badge</sub></sub> | `tests/scenarios/cmd/printf/errors/no_format.yaml` | Duplicate of `basic/no_args.yaml`; also missing bash skip flag |
| 7 | <sub><sub>P2 Badge</sub></sub> | `tests/scenarios/cmd/printf/specifiers/incomplete_with_flags.yaml` | Missing `skip_assert_against_bash: true` — bash errors on incomplete specifiers |
| 8 | <sub><sub>P3 Badge</sub></sub> | `interp/builtins/printf/printf.go:217` | Byte-at-a-time output for literal characters is inefficient |

### `skip_assert_against_bash` Audit

| Scenario file | Bash matches? | Flag needed? | Status |
|--------------|:---:|:---:|--------|
| `shell_features/command_substitution.yaml` | No (shell blocks cmd substitution) | Yes | OK |
| `escapes/octal_single_digit.yaml` | Yes (both produce 0x01) | Unclear | Verify YAML byte encoding |
| `escapes/hex_single_digit.yaml` | Yes (both produce 0x0f) | Unclear | Verify YAML byte encoding |
| `errors/rejected_v_flag.yaml` | No (bash accepts -v) | Yes | OK |
| `specifiers/float_f_upper_nan.yaml` | Yes (bash outputs NAN) | **No** | **Remove flag** |
| `specifiers/b_octal_no_digits.yaml` | Yes (both produce NUL + x) | Unclear | Verify YAML byte encoding |
| `specifiers/b_hex_one_digit.yaml` | Yes (both produce 0x0f) | Unclear | Verify YAML byte encoding |

### Coverage Summary

| Code path | Scenario test | Go test | Status |
|-----------|:---:|:---:|--------|
| Basic format specifiers (%s, %d, etc.) | 10+ scenarios | 20+ tests | **Adequate** |
| Width/precision modifiers | 14 scenarios | 8 tests | **Adequate** |
| Escape sequences | 12 scenarios | 12 tests | **Adequate** |
| %b specifier with escapes | 6 scenarios | 10 tests | **Adequate** |
| Error handling (invalid numbers) | 6 scenarios | 12 tests | **Adequate** |
| Security: %n rejection | 1 scenario | 2 tests | **Adequate** |
| Security: -v rejection | 1 scenario | 2 tests | **Adequate** |
| Security: width/precision clamping | — | 4 tests | **Adequate** |
| Security: iteration limit | — | 1 test | **Adequate** |
| Security: context cancellation | — | 2 tests | **Adequate** |
| Pentest: integer edge cases | — | 12 tests | **Adequate** |
| Pentest: flag injection | — | 6 tests | **Adequate** |
| Incomplete specifier error path | 2 scenarios (wrong expect) | 1 test (wrong expect) | **Needs fix** |
| Unknown specifier error path | — | 1 test (wrong expect) | **Needs fix** |

### Positive Observations

- Excellent security hardening: `%n` rejection, `-v` rejection, width/precision clamping to 10,000, iteration cap at 10,000, context cancellation
- Comprehensive test pyramid: 80+ unit tests, 25 GNU compat tests, 41 pentest tests, 46 YAML scenario tests
- Clean import allowlist: only `context`, `fmt`, `math`, `strconv`, `strings` — all safe, pure functions
- `NoFlags` mode is correct for printf since args can look like flags
- `processBEscapes` properly handles `\c` to stop all output
- Good boundary testing: MaxInt64+1 overflow, huge widths, many-arg iteration limits
- No filesystem access, no stdin reading — purely computes output from args

**Comment by AlexandreYang** [comment]
@datadog @codex make a comprehensive code and security reviews

File: interp/builtins/printf/printf.go [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+			callCtx.Errf("printf: -v: not supported in restricted shell\n")
+			return builtins.Result{Code: 1}
+		case "--":
+			args = args[1:] // skip --
+		}
+	}
+
+	if len(args) == 0 {
+		callCtx.Errf("printf: usage: printf [-v var] format [arguments]\n")
+		return builtins.Result{Code: 1}
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: [Claude Opus 4.6] Done — changed exit code from 1 to 2 for missing format, matching bash behavior. Updated `basic/no_args.yaml` to expect exit code 2. Removed duplicate `errors/no_format.yaml`.

File: interp/builtins/printf/printf.go [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+
+func run(ctx context.Context, callCtx *builtins.CallContext, args []string) builtins.Result {
+	// Manual flag handling: only --help/-h is accepted; -v is rejected.
+	// -- terminates options (allows format strings starting with -).
+	if len(args) > 0 {
+		switch args[0] {
+		case "--help", "-h":
+			callCtx.Out("Usage: printf FORMAT [ARGUMENT]...\n")
+			callCtx.Out("Write formatted output to standard output.\n")
+			return builtins.Result{}
```
Comment (AlexandreYang) [inline interp/builtins/printf/printf.go]: <sub><sub>P2 Badge</sub></sub> **`-h` is not a valid bash printf flag — diverges from bash behavior**

Bash's `printf` does not accept `-h`:
```bash
$ printf -h
bash: printf: -h: invalid option
printf: usage: printf [-v var] format [arguments]
# exit code: 2
```

The implementation treats `-h` as `--help` and returns exit 0, which is an unintentional bash divergence. In bash, `-h` would be rejected as an invalid option.

Since `printf` uses `NoFlags` (meaning unknown flags are passed through as arguments), the simplest approach consistent with bash is to remove `-h` handling entirely. With `NoFlags`, `-h` would be treated as the format string (printing `-h` literally), which is closer to how bash handles unknown single-character flags for printf.

**Remediation**: Remove `"-h"` from the switch case. Only `--help` should show help.

```suggestion
		case "--help":
			callCtx.Out("Usage: printf FORMAT [ARGUMENT]...\n")
			callCtx.Out("Write formatted output to standard output.\n")
			return builtins.Result{Code: 2}
```

File: interp/builtins/printf/printf.go [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+		} else {
+			// Empty argument produces a NUL byte (bash behavior).
+			callCtx.Out("\x00")
+		}
+
+	case 'd', 'i':
+		arg := getStringArg(args, argIdx)
+		val, err := parseIntArg(arg)
+		if err != nil && arg != "" {
+			callCtx.Errf("printf: %s: invalid number\n", arg)
```
Comment (AlexandreYang) [inline interp/builtins/printf/printf.go]: <sub><sub>P3 Badge</sub></sub> **Error message format differs from bash — missing quotes around invalid value**

Bash wraps the invalid value in single quotes:
```
bash: printf: 'abc': invalid number
```

This implementation omits the quotes:
```
printf: abc: invalid number
```

This applies to all `"printf: %s: invalid number\n"` format strings throughout the file.

Note: The scenario test `errors/invalid_number.yaml` uses `stderr_contains` which won't catch this difference. Running `RSHELL_BASH_TEST=1` bash comparison tests may flag this.

**Remediation**: Change `"printf: %s: invalid number\n"` to `"printf: '%s': invalid number\n"` at all occurrences.

**Comment by AlexandreYang** [comment]
@datadog @codex make a comprehensive code and security reviews

**Comment by AlexandreYang** [comment]
**Iteration 1 self-review result: COMMENT (needs fixes)**

- **P1**: 2 findings (must fix)
  - `--` after format string is wrongly stripped
  - `strings.ReplaceAll` missing from allowed symbols
- **P2**: 2 findings (should fix)
  - `-h` is not a valid bash printf flag
  - `--help` exits 0 but bash exits 2
- **P3**: 1 finding (nice to have)
  - Error messages lack quotes around invalid values

Total: 5 findings. Proceeding to fix.

File: interp/builtins/printf/printf.go [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+		val, err := parseFloatArg(arg)
+		if err != nil && arg != "" {
+			callCtx.Errf("printf: '%s': invalid number\n", arg)
+			val = 0
+			goFmt.WriteByte('E')
+			callCtx.Out(fmt.Sprintf(goFmt.String(), val))
+			return false, i, true
+		}
+		goFmt.WriteByte('E')
+		callCtx.Out(fmt.Sprintf(goFmt.String(), val))
```
Comment (AlexandreYang) [inline interp/builtins/printf/printf.go]: <sub><sub>P2 Badge</sub></sub> **Uppercase float format verbs produce wrong Inf/NaN output**

**Category**: Bash Compatibility

The uppercase format verbs `%E`, `%F`, and `%G` produce incorrect output for infinity and NaN values because they don't apply case normalization like the lowercase verbs do.

**Evidence** (verified against `bash` on macOS):

| Input | Expected (bash) | Actual (rshell) |
|-------|-----------------|------------------|
| `printf "%E\n" inf` | `INF` | `+Inf` |
| `printf "%G\n" inf` | `INF` | `+Inf` |
| `printf "%F\n" inf` | `INF` | `+INF` |
| `printf "%E\n" nan` | `NAN` | `NaN` |
| `printf "%G\n" nan` | `NAN` | `NaN` |
| `printf "%F\n" -inf` | `-INF` | `-INF` (correct) |

Go's `fmt.Sprintf` outputs `+Inf`, `-Inf`, and `NaN` for all format verbs. For lowercase verbs (`%e`, `%f`, `%g`), `bashFloat()` correctly normalizes these. For uppercase verbs, no normalization is applied.

**Root cause**: The `%E` case (line 485-489) and `%G` case (line 540-544) don't call `bashFloat()` and then uppercase the result. The `%F` case (line 510-518) uses `strings.ToUpper` on the raw Go output, which preserves the unwanted `+` prefix.

**Remediation**: Create a `bashFloatUpper()` helper that normalizes and uppercases:
```go
func bashFloatUpper(s string) string {
    s = strings.ReplaceAll(s, "NaN", "NAN")
    s = strings.ReplaceAll(s, "+Inf", "INF")
    s = strings.ReplaceAll(s, "-Inf", "-INF")
    s = strings.ReplaceAll(s, "Inf", "INF")
    return s
}
```
Then use `bashFloatUpper()` in the `%E`, `%F`, and `%G` cases.

File: interp/builtins/printf/printf.go [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+		case "--help":
+			callCtx.Errf("printf: usage: printf [-v var] format [arguments]\n")
+			return builtins.Result{Code: 2}
+		case "-v":
+			callCtx.Errf("printf: -v: not supported in restricted shell\n")
+			return builtins.Result{Code: 1}
+		case "--":
+			args = args[1:] // skip --
+		}
+	}
```
Comment (AlexandreYang) [inline interp/builtins/printf/printf.go]: <sub><sub>P2 Badge</sub></sub> **Unknown single-dash flags silently treated as format strings**

**Category**: Bash Compatibility

Bash's `printf` rejects unknown single-dash options with an error. Our implementation falls through and treats them as format strings.

**Evidence**:
```bash
$ bash -c 'printf -h 2>&1; echo exit:$?'
bash: line 1: printf: -h: invalid option
printf: usage: printf [-v var] format [arguments]
exit:2

$ bash -c 'printf -f 2>&1; echo exit:$?'
bash: line 1: printf: -f: invalid option
printf: usage: printf [-v var] format [arguments]
exit:2
```

Our code outputs `-h` / `-f` as literal text with exit 0.

Note: `TestPrintfHelpShort` documents this as intentional with the comment "// -h is not a valid flag in bash; it's treated as a format string" — but that comment is factually incorrect; bash does reject `-h`.

**Remediation**: After the `switch` on `args[0]`, add a check for any arg starting with `-` (but not `--` or `-` alone) to reject it as an invalid option, matching bash behavior:
```go
if len(args[0]) >= 2 && args[0][0] == '-' && args[0][1] != '-' {
    callCtx.Errf("printf: %s: invalid option\n", args[0])
    callCtx.Errf("printf: usage: printf [-v var] format [arguments]\n")
    return builtins.Result{Code: 2}
}
```

**Comment by AlexandreYang** [comment]
@datadog @codex make a comprehensive code and security reviews (iteration 2)

**Comment by AlexandreYang** [comment]
**Iteration 2 self-review result: COMMENT (needs fixes)**

- **P2**: 2 findings (should fix)
  - Uppercase float format verbs (%E, %F, %G) produce wrong Inf/NaN output
  - Unknown single-dash flags treated as format strings instead of invalid option error

Total: 2 findings. Proceeding to fix.

File: interp/builtins/printf/printf.go [hunk 6]
```diff
@@ -0,0 +1,791 @@
...
```
Comment (AlexandreYang) [inline interp/builtins/printf/printf.go]: <sub><sub>P1 Badge</sub></sub> **Unknown flags silently treated as format string instead of rejected**

Bash rejects ALL unknown single-letter flags (e.g. `-h`, `-f`, `-z`) and unknown long flags (e.g. `--follow`, `--foo`) with exit code 2:

```
$ printf -h
bash: printf: -h: invalid option
printf: usage: printf [-v var] format [arguments]
$ printf --follow "%s" hello
bash: printf: --: invalid option
printf: usage: printf [-v var] format [arguments]
```

The current implementation only matches `--help`, `-v`, and `--`, then falls through to treating everything else as the format string. This means `printf -h` outputs `-h` (exit 0) instead of erroring (exit 2).

Fix: after the `--` check, add a catch-all for any arg starting with `-` that has length > 1:

```suggestion
		case "--help":
			callCtx.Errf("printf: usage: printf [-v var] format [arguments]\n")
			return builtins.Result{Code: 2}
		case "-v":
			callCtx.Errf("printf: -v: not supported in restricted shell\n")
			return builtins.Result{Code: 1}
		case "--":
			args = args[1:] // skip --
		default:
			if len(args[0]) > 1 && args[0][0] == '-' {
				callCtx.Errf("printf: %s: invalid option\n", args[0])
				callCtx.Errf("printf: usage: printf [-v var] format [arguments]\n")
				return builtins.Result{Code: 2}
			}
```

Note: the existing `TestPrintfHelpShort` test asserts the wrong behavior (`-h` treated as format string). That test should be updated to expect exit 2 and an error. The `TestPentestUnknownFlags` and `TestPentestFollowFlag` tests also encode the wrong behavior.

**Review by AlexandreYang — COMMENTED** [review]
## Review Summary

Reviewed the full printf builtin implementation including `printf.go`, tests (`printf_test.go`, `printf_pentest_test.go`, `printf_gnu_compat_test.go`), 85+ YAML scenario tests, `SHELL_FEATURES.md` documentation, registration in `register_builtins.go`, and allowed-symbols additions.

**Overall assessment: needs fixes** — one P1 bash compatibility bug in flag handling, two P2 bash compatibility issues.

### Security

The implementation is security-sound:
- No filesystem or stdin access — pure format+args processing
- `%n` correctly rejected (security risk)
- `-v` correctly rejected (restricted shell)
- Width/precision clamped to 10,000 to prevent OOM
- Format reuse loop bounded to 10,000 iterations with context cancellation
- All imports are on the allowed-symbols list (only `context`, `fmt`, `math`, `strconv`, `strings`)
- `fmt.Sprintf` format strings are safely constructed from validated components (flags limited to `-+' '0#`, width/precision are digits, verb is a known letter)

### Test Coverage

Excellent coverage overall: 85+ YAML scenario tests, comprehensive Go unit tests, GNU compat tests, and pentest tests covering integer edge cases, resource exhaustion, flag injection, etc.

### Findings

| # | Priority | File | Finding |
|---|----------|------|--------|
| 1 | <sub><sub>P1 Badge</sub></sub> | `interp/builtins/printf/printf.go:142` | Unknown flags (`-h`, `-f`, etc.) silently treated as format string instead of rejected |
| 2 | <sub><sub>P2 Badge</sub></sub> | `interp/builtins/printf/printf.go:616` | Bare quote character constant (`'`) errors instead of returning 0 |
| 3 | <sub><sub>P2 Badge</sub></sub> | `interp/builtins/printf/printf.go:573` | Unknown format specifier continues processing instead of stopping |

### Positive Observations

- Clean separation of concerns: format string processing, escape handling, and specifier formatting are well-structured
- Bash float casing (`bashFloat`/`bashFloatUpper`) is a nice touch for NaN/Inf compatibility
- The `NoFlags` approach is correct for printf since format strings can start with `-`
- Width/precision clamping and iteration bounding are solid DoS protections
- Comprehensive YAML scenario coverage with proper `skip_assert_against_bash: true` only for intentional divergences (`-v`, `%n`)
- GNU compat test suite provides additional byte-for-byte verification

File: interp/builtins/printf/printf.go [hunk 7]
```diff
@@ -0,0 +1,778 @@
...
+		return 0, nil
+	}
+
+	// Character constant.
+	if len(s) >= 2 && (s[0] == '\'' || s[0] == '"') {
+		return float64(s[1]), nil
+	}
+
+	// Handle hex integers used as float args (0xff etc).
+	if len(s) > 2 && s[0] == '0' && (s[1] == 'x' || s[1] == 'X') {
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: Fixed in commit 02be117. `parseFloatArg` now strips the sign prefix before checking for hex/octal prefixes, so `-0xff` is correctly parsed as `-255.000000`. Added scenario test `numeric/signed_hex_float.yaml`.

File: interp/builtins/printf/printf.go [hunk 8]
```diff
@@ -0,0 +1,778 @@
...
+
+	case 'c':
+		arg := getStringArg(args, argIdx)
+		if len(arg) > 0 {
+			// %c prints the first character (byte).
+			goFmt.WriteByte('c')
+			callCtx.Out(fmt.Sprintf(goFmt.String(), arg[0]))
+		} else {
+			// Empty argument produces a NUL byte (bash behavior).
+			callCtx.Out("\x00")
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: Already working correctly. The code uses `goFmt.WriteByte('c')` and `fmt.Sprintf(goFmt.String(), ch)` where `ch` is 0 for empty args. Go's `%3c` with byte 0 correctly outputs 2 spaces + NUL, matching bash. Verified with `bash -c 'printf "%3c" "" | xxd'` which shows identical output.

File: interp/builtins/printf/printf.go [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+// hex (0x-prefix), and character constants ('X or "X).
+func parseIntArg(s string) (int64, error) {
+	if s == "" {
+		return 0, nil
+	}
+
+	// Character constant: 'X or "X
+	if len(s) >= 2 && (s[0] == '\'' || s[0] == '"') {
+		return int64(s[1]), nil
+	}
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: Fixed in commit 02be117. All three parse functions (`parseIntArg`, `parseUintArg`, `parseFloatArg`) now handle bare quote characters (length 1) by returning 0 instead of falling through to the integer parser. Added scenario test `numeric/bare_quote_char_constant.yaml`.

File: interp/builtins/printf/printf.go [hunk 10]
```diff
@@ -0,0 +1,764 @@
...
+	i++ // consume verb
+
+	// Build Go format string.
+	var goFmt strings.Builder
+	goFmt.WriteByte('%')
+	goFmt.WriteString(flags.String())
+	goFmt.WriteString(width)
+	if hasPrecision {
+		goFmt.WriteByte('.')
+		goFmt.WriteString(precision)
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: [Claude Opus 4.6] Already fixed in a previous commit. Lines 348-350 handle this: when precision from `*` is negative, `hasPrecision` is set to `false`, treating it as unspecified. `printf "%.*s\n" -2 abc` correctly outputs `abc`.

File: interp/builtins/printf/printf.go [hunk 11]
```diff
@@ -0,0 +1,785 @@
...
+	}
+
+	format := args[0]
+	fmtArgs := args[1:]
+
+	// Strip a leading "--" from format arguments (allows negative numbers
+	// after the format string: printf "%d" -- -42).
+	if len(fmtArgs) > 0 && fmtArgs[0] == "--" {
+		fmtArgs = fmtArgs[1:]
+	}
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: [Claude Opus 4.6] Already fixed in a previous commit. The post-format `--` stripping code was removed. `--` is only handled at line 150 (before the format string). After the format string, `--` is treated as a data argument, matching bash behavior.

File: interp/builtins/printf/printf.go (hunk 3 above)
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: [Claude Opus 4.6] Already fixed in a previous commit. All `invalid number` error messages now wrap the value in single quotes: `printf: '%s': invalid number`. This matches bash's format `bash: printf: 'abc': invalid number`.

File: interp/builtins/printf/printf.go [hunk 12]
```diff
@@ -0,0 +1,826 @@
...
+		return false, i, true
+	}
+
+	verb := s[i]
+	i++ // consume verb
+
+	// Build Go format string.
+	var goFmt strings.Builder
+	goFmt.WriteByte('%')
+	goFmt.WriteString(flags.String())
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: [Claude Opus 4.6] Already addressed — the `stripSignFlags` function was added in a prior commit (748a74f) and is applied for `%o`, `%u`, `%x`, `%X` at the format string construction stage (lines 401-403). Verified with `docker run --rm debian:bookworm-slim bash -c 'printf "%+u\n" 1; printf "% x\n" 255'` — both output `1` and `ff` respectively, matching our implementation.

**Review by AlexandreYang — COMMENTED** [review]
## Review Summary

Reviewed the `printf` builtin implementation — a new 990-line builtin that implements POSIX/bash printf formatting with format specifiers (%s, %b, %c, %d, %i, %o, %u, %x, %X, %e, %E, %f, %F, %g, %G), escape sequences, width/precision, star arguments, format reuse, and big.Int exact formatting for large integers.

**Overall assessment: needs fixes** — the `TestBuiltinAllowedSymbols` test fails because several symbols used by the new builtin are not registered in the allowlist. This must be fixed before merge. No security issues found.

### Findings Summary

| # | Priority | File | Finding |
|---|----------|------|--------|
| 1 | <sub><sub>P1 Badge</sub></sub> | `tests/allowed_symbols_test.go` | Missing allowlist entries: `math/big` package, `strings.ContainsRune`, `strings.Repeat` cause test failure |
| 2 | <sub><sub>P3 Badge</sub></sub> | `tests/scenarios/cmd/printf/escapes/hex_single_digit.yaml` | Unnecessary `skip_assert_against_bash: true` — behavior matches bash |
| 3 | <sub><sub>P3 Badge</sub></sub> | `tests/scenarios/cmd/printf/escapes/octal_single_digit.yaml` | Unnecessary `skip_assert_against_bash: true` — behavior matches bash |

### Positive Observations

- **Security**: `%n` is explicitly rejected with a clear error message. `-v` flag is blocked. No filesystem or stdin access. Format reuse is bounded to 10,000 iterations with context cancellation checks. Width/precision are clamped to 10,000 to prevent memory exhaustion. Good.
- **Bash compatibility**: Comprehensive handling of edge cases — NaN/Inf casing, unsigned negative wrapping, character constants, hex/octal input parsing, sign flag stripping on unsigned conversions, big.Int exact formatting for large integers with %f/%F.
- **Test coverage**: Excellent — 830+ lines of Go tests (functional, GNU compat, pentest) plus 80+ YAML scenario tests covering all specifiers, escapes, width/precision, error cases, shell integration, and edge cases.
- **Code quality**: Well-structured with clear separation between format parsing, specifier handling, escape processing, and number parsing. Good documentation in the package comment.

File: tests/allowed_symbols_test.go [hunk 13]
```diff
@@ -68,6 +68,8 @@ var builtinAllowedSymbols = []string{
 	"io.ReadCloser",
 	// io.Reader — interface type; no side effects.
 	"io.Reader",
+	// math.Inf — returns positive or negative infinity; pure function, no I/O.
+	"math.Inf",
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply tests/allowed_symbols_test.go]: [Claude Opus 4.6] Done — added `math/big.Int`, `math/big.NewInt`, `strings.ContainsRune`, and `strings.Repeat` to the builtin allowed symbols list. `TestBuiltinAllowedSymbols` passes now.

File: tests/scenarios/cmd/printf/escapes/hex_single_digit.yaml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+description: Single hex digit escape in format string.
+input:
+  script: |+
+    printf '\xF'
+expect:
+  stdout: "\x0f"
+  exit_code: 0
+skip_assert_against_bash: true
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply tests/scenarios/cmd/printf/escapes/hex_single_digit.yaml]: [Claude Opus 4.6] Done — removed `skip_assert_against_bash: true` from this test. Bash comparison passes.

**Comment by AlexandreYang** [comment]
@datadog @codex make a comprehensive code and security reviews (iteration 4)

File: interp/builtins/printf/printf.go [hunk 15]
```diff
@@ -0,0 +1,785 @@
...
+		// \0NNN — octal (0 + up to 3 digits)
+		val, consumed := parseOctal(s[2:], 3)
+		return string(rune(val)), 2 + consumed
+	case 'x':
+		// \xHH — hex (up to 2 digits)
+		val, consumed := parseHex(s[2:], 2)
+		if consumed == 0 {
+			return "\\x", 2
+		}
+		return string(rune(val)), 2 + consumed
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/printf/printf.go]: [Claude Opus 4.6] Already fixed in a previous commit. All escape sequence handlers (`\0NNN`, `\xHH`, `\NNN`) use `string([]byte{byte(val)})` to emit raw bytes instead of `string(rune(val))`. Values above 0x7F are emitted as single bytes, not UTF-8 encoded.
